File: python/src/arker/build_spec.py
"""Parse a Dockerfile into the instruction list the SDK build driver walks.

Text in, instructions out — no network, no VM, no filesystem. The driver in
``build.py`` is what executes them: ``FROM`` becomes a real image fork, ``RUN``
becomes ``vm.run(...)``, ``COPY`` becomes a directory sync into the VM.

## Why this lives in the SDK and not on the server

Every instruction except ``COPY`` is self-contained text that a server could
interpret perfectly well — and one used to. ``COPY`` is different: its sources
are files on the machine running this code. A server-side build would need the
client to upload a build context first, which means an upload endpoint, object
storage, context ids, and a size ceiling on however big someone's project is.

Driving the build from here removes all of that. ``COPY`` becomes
``vm.sync_dir(...)``, which already streams a tarball into the guest and, being
manifest-diffed, ships only the files that actually changed.

## Scope

Supported: ``FROM``, ``RUN``, ``COPY``, ``ADD <url>``, ``ENV``, ``WORKDIR``,
``USER``, ``ARG``, ``LABEL``, ``EXPOSE``, ``ENTRYPOINT``, ``CMD``.

Refused by name, rather than silently dropped:

* **Multi-stage builds** — more than one ``FROM``, or a ``COPY --from=``.
  Resolving stage aliases and deciding what a final stage means for a single
  long-lived VM is its own feature.
* **``ARG``-substituted ``FROM``** — ``FROM`` resolves through a real image
  fork, so it has to be a literal reference.
* **Everything else** (``VOLUME``, ``ONBUILD``, ``HEALTHCHECK``, ``SHELL``,
  ``STOPSIGNAL``, ``MAINTAINER``, ...) — named in the error so an unsupported
  directive never passes silently.

``${VAR}`` substitution into instruction text is not performed. A ``RUN`` sees
``$VAR`` as an ordinary shell variable because the shell expands it; a
``WORKDIR /app-$VERSION`` does not, and is left as written.
"""

from __future__ import annotations

import io
import json
import re
import shlex
from collections.abc import Iterator
from dataclasses import dataclass, field

from dockerfile_parse import DockerfileParser

__all__ = [
    "Add",
    "Arg",
    "Cmd",
    "Copy",
    "DockerfileError",
    "Entrypoint",
    "Env",
    "Expose",
    "Label",
    "ParsedDockerfile",
    "Run",
    "Step",
    "User",
    "Workdir",
    "parse_dockerfile",
]


class DockerfileError(ValueError):
    """A Dockerfile this SDK will not build, with the reason named."""


@dataclass(frozen=True)
class Run:
    """``RUN <cmd>`` — executed against the VM via ``vm.run``."""

    command: str


@dataclass(frozen=True)
class Copy:
    """``COPY <src>... <dest>`` — synced from the build context into the VM.

    ``sources`` are as written, relative to the context root; the driver
    resolves and globs them locally. ``chown`` is carried rather than dropped
    so the driver can apply it after the sync.
    """

    sources: list[str]
    dest: str
    chown: str | None = None


@dataclass(frozen=True)
class Add:
    """``ADD <url> <dest>`` — fetched inside the guest.

    Only the URL form is supported. A local-source ``ADD`` differs from
    ``COPY`` mainly by auto-extracting archives, which this does not do, so it
    is refused rather than silently behaving like ``COPY``.
    """

    url: str
    dest: str
    checksum: str | None = None


@dataclass(frozen=True)
class Env:
    """``ENV k=v [k=v ...]`` — exported for later ``RUN``s and persisted."""

    pairs: list[tuple[str, str]]


@dataclass(frozen=True)
class Workdir:
    """``WORKDIR <dir>`` — created if missing, then ``cd``'d into."""

    path: str


@dataclass(frozen=True)
class User:
    """``USER <name>`` — subsequent ``RUN``s execute as this user."""

    name: str


@dataclass(frozen=True)
class Arg:
    """``ARG name[=default]``.

    There is no ``--build-arg`` channel, so a declared default is the only
    value it can take; one without a default is inert, matching Docker's own
    unset-without-override behaviour.
    """

    name: str
    default: str | None = None


@dataclass(frozen=True)
class Label:
    """``LABEL k=v`` — recorded, no runtime effect on a VM."""

    pairs: list[tuple[str, str]]


@dataclass(frozen=True)
class Expose:
    """``EXPOSE <port>`` — recorded; VMs are reached via run/SSH, not -p."""

    port: str


@dataclass(frozen=True)
class Entrypoint:
    """``ENTRYPOINT`` — composed with ``CMD`` by the driver."""

    command: str


@dataclass(frozen=True)
class Cmd:
    """``CMD`` — arguments to ``ENTRYPOINT``, or the whole command alone."""

    command: str


Step = Run | Copy | Add | Env | Workdir | User | Arg | Label | Expose | Entrypoint | Cmd


@dataclass
class ParsedDockerfile:
    """A Dockerfile that passed validation."""

    base_image: str
    steps: list[Step] = field(default_factory=list)


#: Directives that carry no execution semantics for a VM but are accepted so a
#: real-world Dockerfile parses. They are recorded, never silently dropped.
_INERT = {"LABEL", "EXPOSE"}

#: Everything this SDK knows about. Anything else is named in the error.
_KNOWN = {
    "FROM",
    "RUN",
    "COPY",
    "ADD",
    "ENV",
    "WORKDIR",
    "USER",
    "ARG",
    "ENTRYPOINT",
    "CMD",
} | _INERT


def _instructions(text: str) -> Iterator[tuple[str, str]]:
    """Yield ``(DIRECTIVE, argument)`` pairs from Dockerfile text.

    Lexing is delegated to ``dockerfile-parse`` rather than hand-rolled. That
    layer — line continuations, comments (including a ``#`` line *inside* a
    continued instruction), the ``# escape=`` parser directive, CRLF — is
    exactly where a hand-written splitter goes quietly wrong, and it is the
    part this library does well. It hands back the instruction NAME and its
    raw argument string; parsing the arguments (flags, exec-form JSON,
    ``key=value``) is still ours, because the library does not go that deep.

    It is also what Harbor and every sandbox provider it ships (e2b, modal,
    runloop, novita, blaxel, beam, …) use for this same job, and Arker is
    itself a Harbor provider.
    """
    parser = DockerfileParser(fileobj=io.BytesIO())
    parser.content = text
    for instruction in parser.structure:
        directive = str(instruction.get("instruction", "")).upper()
        if directive == "COMMENT":
            continue
        yield directive, str(instruction.get("value", "")).strip()


def _exec_form(argument: str) -> list[str] | None:
    """Return the argv of an exec-form instruction, or None if shell-form."""
    if not argument.startswith("["):
        return None
    try:
        parsed = json.loads(argument)
    except ValueError:
        return None
    if isinstance(parsed, list) and all(isinstance(item, str) for item in parsed):
        return parsed
    return None


def _command_line(argument: str) -> str:
    """Render RUN/ENTRYPOINT/CMD as one shell command line.

    Exec form does not go through a shell in Docker, but the only execution
    primitive we have is "a command line in a shell", so each element is quoted
    and joined — which invokes the same argv a direct exec would.
    """
    argv = _exec_form(argument)
    if argv is None:
        return argument
    return " ".join(shlex.quote(item) for item in argv)


def _validate_env_key(key: str, directive: str) -> str:
    """A key is a bare token in `export k=v` and CANNOT be quoted.

    So it is the one interpolation in the whole driver that quoting cannot
    make safe, and it has to be validated instead. `ENV a;id;b=1` parses
    happily and would emit `export a;id;b=1`. Docker refuses the same input
    with "invalid environment variable name".
    """
    if not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", key):
        raise DockerfileError(
            f"{directive} name {key!r} is not a valid environment variable name "
            "(letters, digits and underscore, not starting with a digit)"
        )
    return key


def _key_value_pairs(argument: str, directive: str) -> list[tuple[str, str]]:
    """Parse ``k=v k=v`` and the legacy ``ENV key value`` single-pair form."""
    try:
        tokens = shlex.split(argument)
    except ValueError as error:
        raise DockerfileError(f"{directive} is not parseable: {error}") from error
    if not tokens:
        raise DockerfileError(f"{directive} requires at least one key")
    if "=" not in tokens[0]:
        # Legacy `ENV key value with spaces` — everything after the key is the
        # value, so it cannot be split into further pairs.
        if len(tokens) < 2:
            raise DockerfileError(f"{directive} {tokens[0]} has no value")
        return [(_validate_env_key(tokens[0], directive), " ".join(tokens[1:]))]
    pairs = []
    for token in tokens:
        key, sep, value = token.partition("=")
        if not sep or not key:
            raise DockerfileError(f"{directive} expects key=value, got: {token}")
        pairs.append((_validate_env_key(key, directive), value))
    return pairs


def _split_flags(argument: str, directive: str) -> tuple[dict[str, str], list[str]]:
    """Peel leading ``--flag=value`` tokens off an instruction's arguments."""
    try:
        tokens = shlex.split(argument)
    except ValueError as error:
        raise DockerfileError(f"{directive} is not parseable: {error}") from error
    flags: dict[str, str] = {}
    while tokens and tokens[0].startswith("--"):
        name, _, value = tokens.pop(0)[2:].partition("=")
        flags[name] = value
    return flags, tokens


def parse_dockerfile(text: str) -> ParsedDockerfile:
    """Parse and validate ``text``, or raise :class:`DockerfileError`."""
    base_image: str | None = None
    steps: list[Step] = []
    from_count = 0

    for directive, argument in _instructions(text):
        if directive not in _KNOWN:
            raise DockerfileError(
                f"`{directive}` is not supported; supported directives are "
                f"{', '.join(sorted(_KNOWN))} (single-stage only)"
            )

        if directive == "FROM":
            from_count += 1
            if from_count > 1:
                raise DockerfileError(
                    "multi-stage dockerfiles are not supported (found more than one "
                    "FROM): flatten to a single stage, or fork from an already-built "
                    "image reference"
                )
            image = argument.split(" AS ")[0].split(" as ")[0].strip()
            if "$" in image:
                raise DockerfileError(
                    f"FROM referencing a build ARG (`{image}`) is not supported: give a "
                    "literal base image reference, for example `FROM ubuntu:24.04`"
                )
            if not image:
                raise DockerfileError("FROM requires an image reference")
            base_image = image

        elif directive == "RUN":
            if not argument:
                raise DockerfileError("RUN requires a command")
            steps.append(Run(_command_line(argument)))

        elif directive == "COPY":
            flags, tokens = _split_flags(argument, "COPY")
            if "from" in flags:
                raise DockerfileError(
                    f"COPY --from={flags['from']} is not supported: cross-stage copies "
                    "need multi-stage builds, which this path does not implement"
                )
            if len(tokens) < 2:
                raise DockerfileError(f"COPY requires at least one source and a destination, got: {argument}")
            steps.append(Copy(tokens[:-1], tokens[-1], chown=flags.get("chown")))

        elif directive == "ADD":
            flags, tokens = _split_flags(argument, "ADD")
            if len(tokens) != 2:
                raise DockerfileError(
                    f"ADD {argument} is not supported: only `ADD <url> <dest>` is, with a "
                    "single URL source and destination"
                )
            source, dest = tokens
            if not source.startswith(("http://", "https://")):
                raise DockerfileError(
                    f"ADD {source} is not supported: only a URL source is. Use COPY for "
                    "local files — note ADD's archive auto-extraction is not implemented, "
                    "which is why a local ADD is refused rather than treated as COPY"
                )
            steps.append(Add(source, dest, checksum=flags.get("checksum")))

        elif directive == "ENV":
            steps.append(Env(_key_value_pairs(argument, "ENV")))

        elif directive == "WORKDIR":
            if not argument:
                raise DockerfileError("WORKDIR requires a directory")
            steps.append(Workdir(argument))

        elif directive == "USER":
            if not argument:
                raise DockerfileError("USER requires a username")
            steps.append(User(argument))

        elif directive == "ARG":
            name, sep, default = argument.partition("=")
            name = name.strip()
            if not name:
                raise DockerfileError("ARG requires a name")
            _validate_env_key(name, "ARG")
            steps.append(Arg(name, default.strip() if sep else None))

        elif directive == "LABEL":
            steps.append(Label(_key_value_pairs(argument, "LABEL")))

        elif directive == "EXPOSE":
            if not argument:
                raise DockerfileError("EXPOSE requires a port")
            steps.append(Expose(argument))

        elif directive == "ENTRYPOINT":
            steps.append(Entrypoint(_command_line(argument)))

        elif directive == "CMD":
            steps.append(Cmd(_command_line(argument)))

    if base_image is None:
        raise DockerfileError("dockerfile has no FROM instruction")
    return ParsedDockerfile(base_image=base_image, steps=steps)

File: typescript/src/buildSpec.ts
/**
 * Parse a Dockerfile into the instruction list the build driver walks.
 *
 * Text in, instructions out. No network, no VM, no filesystem. `build.ts`
 * executes them: `FROM` becomes a real image fork, `RUN` becomes `vm.run`,
 * `COPY` becomes a sync into the VM.
 *
 * ## Why this is in the SDK
 *
 * Every instruction except `COPY` is self-contained text a server could
 * interpret, and one used to. `COPY` is different: its sources are files on the
 * machine running this code. A server-side build would need the client to
 * upload a build context first, which means an upload endpoint, object storage,
 * and a ceiling on how large a project may be. Building from here turns `COPY`
 * into `vm.syncDir`, which streams into the guest and re-sends only what
 * changed.
 *
 * ## Scope
 *
 * Supported: `FROM`, `RUN`, `COPY`, `ADD <url>`, `ENV`, `WORKDIR`, `USER`,
 * `ARG`, `LABEL`, `EXPOSE`, `ENTRYPOINT`, `CMD`, `SHELL`.
 *
 * `SHELL` replaces the interpreter for the shell form of later `RUN`,
 * `ENTRYPOINT` and `CMD`, as it does in Docker. It is state, not a step:
 * nothing executes it. Exec form is left alone throughout, because it does not
 * go through an interpreter at all.
 *
 * Refused by name rather than silently dropped: multi-stage builds (more than
 * one `FROM`, or a `COPY --from=`), an `ARG`-substituted `FROM` (it resolves
 * through a real image fork, so it must be literal), a local-source `ADD` (its
 * one advantage over `COPY` is archive auto-extraction, which we do not do),
 * and every other directive.
 *
 * `${VAR}` substitution into instruction text is not performed. A `RUN` sees
 * `$VAR` because the shell expands it; `WORKDIR /app-$VERSION` does not.
 */

import { Copy as AstCopy, DockerfileParser, From, Instruction } from "dockerfile-ast";

/** A Dockerfile this SDK will not build, with the reason named. */
export class DockerfileError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DockerfileError";
  }
}

/** `RUN <cmd>` — executed against the VM via `vm.run`. */
export type RunStep = { kind: "run"; command: string };
/**
 * `COPY <src>... <dest>` — synced from the build context into the VM.
 * `sources` are as written; the driver resolves and globs them locally.
 */
export type CopyStep = { kind: "copy"; sources: string[]; dest: string; chown?: string };
/** `ADD <url> <dest>` — fetched inside the guest. */
export type AddStep = { kind: "add"; url: string; dest: string; checksum?: string };
/** `ENV k=v` — exported for later RUNs, and persists onto the delivered VM. */
export type EnvStep = { kind: "env"; pairs: [string, string][] };
/** `WORKDIR <dir>` — created if missing, then entered. */
export type WorkdirStep = { kind: "workdir"; path: string };
/** `USER <name>` — subsequent RUNs execute as this user. */
export type UserStep = { kind: "user"; name: string };
/** `ARG name[=default]`; without a default it is inert, as in Docker. */
export type ArgStep = { kind: "arg"; name: string; default?: string };
/** Recorded, but with no runtime effect on a VM. */
export type InertStep = { kind: "label" | "expose" | "entrypoint" | "cmd"; value: string };

export type Step =
  | RunStep
  | CopyStep
  | AddStep
  | EnvStep
  | WorkdirStep
  | UserStep
  | ArgStep
  | InertStep;

export type ParsedDockerfile = { baseImage: string; steps: Step[] };

const KNOWN = new Set([
  "FROM", "RUN", "COPY", "ADD", "ENV", "WORKDIR",
  "USER", "ARG", "LABEL", "EXPOSE", "ENTRYPOINT", "CMD", "SHELL",
]);

/**
 * Quote one argument for a shell command line, the way `shlex.quote` does:
 * only when it needs it, so ordinary tokens stay readable.
 */
function shellQuote(value: string): string {
  if (value.length > 0 && /^[A-Za-z0-9_@%+=:,./-]+$/.test(value)) return value;
  return `'${value.replace(/'/g, `'"'"'`)}'`;
}

/** The argv of an exec-form instruction, or undefined if it is shell form. */
function execForm(raw: string): string[] | undefined {
  const trimmed = raw.trim();
  if (!trimmed.startsWith("[")) return undefined;
  try {
    const argv = JSON.parse(trimmed);
    if (Array.isArray(argv) && argv.every((item) => typeof item === "string")) return argv;
  } catch {
    // Not valid JSON, so it was never exec form: a shell-form command may
    // legitimately start with `[` (the `test` builtin), so this is not an error.
  }
  return undefined;
}

/**
 * Render RUN/ENTRYPOINT/CMD as one command line.
 *
 * Exec form does not go through a shell in Docker, but the only execution
 * primitive here is "a command line in a shell", so each element is quoted and
 * joined. That invokes the same argv a direct exec would, via the shell's own
 * fork+exec.
 *
 * `shell` is the interpreter set by a preceding `SHELL`. Docker applies it to
 * the SHELL FORM of all three instructions, and to exec form of none of them,
 * which is what the branch below encodes. Absent a `SHELL` the argument is
 * passed through untouched, so the guest's own default shell runs it as before.
 */
function commandLine(raw: string, shell: string[] | undefined): string {
  const argv = execForm(raw);
  if (argv !== undefined) return argv.map(shellQuote).join(" ");
  if (shell === undefined) return raw.trim();
  return [...shell, raw.trim()].map(shellQuote).join(" ");
}

function keyValuePairs(instruction: Instruction, directive: string): [string, string][] {
  const properties = (instruction as unknown as {
    getProperties?: () => { getName(): string; getValue(): string | null }[];
  }).getProperties?.();
  if (!properties || properties.length === 0) {
    throw new DockerfileError(`${directive} requires at least one key`);
  }
  return properties.map((property) => [property.getName(), property.getValue() ?? ""]);
}

/** Parse and validate `text`, or throw {@link DockerfileError}. */
export function parseDockerfile(text: string): ParsedDockerfile {
  const instructions = DockerfileParser.parse(text).getInstructions();
  let baseImage: string | undefined;
  let fromCount = 0;
  const steps: Step[] = [];
  let shell: string[] | undefined;

  for (const instruction of instructions) {
    const directive = (instruction.getKeyword() ?? "").toUpperCase();
    const argument = (instruction.getArgumentsContent() ?? "").trim();

    if (!KNOWN.has(directive)) {
      throw new DockerfileError(
        `\`${directive}\` is not supported; supported directives are ` +
          `${[...KNOWN].sort().join(", ")} (single-stage only)`,
      );
    }

    switch (directive) {
      case "FROM": {
        fromCount += 1;
        if (fromCount > 1) {
          throw new DockerfileError(
            "multi-stage dockerfiles are not supported (found more than one FROM): " +
              "flatten to a single stage, or fork from an already-built image reference",
          );
        }
        const image = instruction instanceof From
          ? instruction.getImage() ?? ""
          : argument.split(/\s+AS\s+/i)[0];
        if (image.includes("$")) {
          throw new DockerfileError(
            `FROM referencing a build ARG (\`${image}\`) is not supported: give a ` +
              "literal base image reference, for example `FROM ubuntu:24.04`",
          );
        }
        if (!image) throw new DockerfileError("FROM requires an image reference");
        baseImage = image;
        break;
      }
      case "RUN": {
        if (!argument) throw new DockerfileError("RUN requires a command");
        steps.push({ kind: "run", command: commandLine(argument, shell) });
        break;
      }
      case "COPY": {
        const copy = instruction as AstCopy;
        const flags = copy.getFlags?.() ?? [];
        const from = flags.find((flag) => flag.getName() === "from");
        if (from) {
          throw new DockerfileError(
            `COPY --from=${from.getValue()} is not supported: cross-stage copies need ` +
              "multi-stage builds, which this path does not implement",
          );
        }
        const chown = flags.find((flag) => flag.getName() === "chown")?.getValue();
        const tokens = copy.getArguments().map((a) => a.getValue());
        if (tokens.length < 2) {
          throw new DockerfileError(
            `COPY requires at least one source and a destination, got: ${argument}`,
          );
        }
        steps.push({
          kind: "copy",
          sources: tokens.slice(0, -1),
          dest: tokens[tokens.length - 1]!,
          ...(chown ? { chown } : {}),
        });
        break;
      }
      case "ADD": {
        const addFlags = (instruction as AstCopy).getFlags?.() ?? [];
        const checksum = addFlags.find((flag) => flag.getName() === "checksum")?.getValue();
        const tokens = instruction.getArguments().map((a) => a.getValue());
        if (tokens.length !== 2) {
          throw new DockerfileError(
            `ADD ${argument} is not supported: only \`ADD <url> <dest>\` is, with a ` +
              "single URL source and destination",
          );
        }
        const [source, dest] = tokens as [string, string];
        if (!/^https?:\/\//.test(source)) {
          throw new DockerfileError(
            `ADD ${source} is not supported: only a URL source is. Use COPY for local ` +
              "files. ADD's archive auto-extraction is not implemented, which is why a " +
              "local ADD is refused rather than treated as COPY",
          );
        }
        steps.push({ kind: "add", url: source, dest, ...(checksum ? { checksum } : {}) });
        break;
      }
      case "ENV": {
        steps.push({ kind: "env", pairs: keyValuePairs(instruction, "ENV") });
        break;
      }
      case "ARG": {
        const properties = (instruction as unknown as {
          getProperties(): { getName(): string; getValue(): string | null }[];
        }).getProperties();
        if (properties.length === 0) throw new DockerfileError("ARG requires a name");
        for (const property of properties) {
          const value = property.getValue();
          steps.push({
            kind: "arg",
            name: property.getName(),
            ...(value === null ? {} : { default: value }),
          });
        }
        break;
      }
      case "WORKDIR": {
        if (!argument) throw new DockerfileError("WORKDIR requires a directory");
        steps.push({ kind: "workdir", path: argument });
        break;
      }
      case "USER": {
        if (!argument) throw new DockerfileError("USER requires a username");
        steps.push({ kind: "user", name: argument });
        break;
      }
      case "EXPOSE": {
        if (!argument) throw new DockerfileError("EXPOSE requires a port");
        steps.push({ kind: "expose", value: argument });
        break;
      }
      case "LABEL": {
        steps.push({ kind: "label", value: argument });
        break;
      }
      case "ENTRYPOINT": {
        steps.push({ kind: "entrypoint", value: commandLine(argument, shell) });
        break;
      }
      case "CMD": {
        steps.push({ kind: "cmd", value: commandLine(argument, shell) });
        break;
      }
      case "SHELL": {
        const argv = execForm(argument);
        if (argv === undefined) {
          throw new DockerfileError(
            'SHELL must be given in exec form, for example `SHELL ["/bin/bash", "-c"]`',
          );
        }
        if (argv.length === 0) {
          throw new DockerfileError("SHELL requires at least one element, the interpreter to run");
        }
        shell = argv;
        break;
      }
    }
  }

  if (!baseImage) throw new DockerfileError("dockerfile has no FROM instruction");
  return { baseImage, steps };
}
